Initial review

pzakha opened the pull request:
This is the initial review for linux-pkg.
See http://reviews.delphix.com/r/44755/ for the first round of reviews.
Moving to github due to some issues with reviewboard.

pzakha commented:
@shartse

pzakha commented on README.md:
```diff
@@ -0,0 +1,736 @@
+Regarding the source code, we expect that each package repository and the
+linux-pkg repository itself follows the Delphix branching policy outlined
+[here](https://docs.delphix.com/pages/viewpage.action?spaceKey=RE&title=New+Branching+Mechanism).
+When creating a new branch or release for the Delphix Appliance, an external
```
We do not have a script for the `linux-pkg` packages yet, but I've been told we do have something for other projects such as virtualization, masking, etc. Perhaps @sebroy would know a bit more.

shartse replied:
I was involved in the branching process for 5.3 - it's a pretty manual process

pzakha commented on package-lists/updateall.pkgs:
```diff
@@ -0,0 +1,10 @@
+# but prevent it from being merged, add the package to
+# auto-merge-blacklist.pkg.
+#
+
```
The update hook would fail for the Delphix package and we would see that failure in the job's status report.

shartse commented:
I think I've reasoned through these changes pretty well on a high level, but I feel less confident that I've carefully reviewed all the code. Do you have ideas about how we can add an initial, simple testing framework for this? An easy place to start might be to check error handling when passing mutually exclusive flags. 

The git-hooks gate, for example, just has a shell script that executes a series of python tests and we encourage developers to re-run that script after making changes. I'd feel more confident getting this out there if we have at least the beginning of a test framework.

pzakha commented:
@shartse I could add a minimal `make test` rule for `linux-pkg` that would run some basic tests. Since this framework will be closely integrated with jenkins jobs `linux-pkg-build` and `linux-pkg-update`, I'm also thinking of having them being run as a requirement for any changes to the main scripts.

shartse commented:
Yeah, that sounds like a good idea

prakashsurya commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+	# appliance. This is not necessarily the latest kernel version
+	# available, but it is one that will match our appliance.
+	#
+	if [[ "$(apt-cache show --no-all-versions "linux-image-${platform}" \
```
can you explain the assumptions here, e.g. how the linux-image package work.. I presume there's something like `linux-image-aws` which depends on another package, like `linux-image-$VERSION-aws`.. and we use this dependency to figure out the value of `$VERSION`, which is what we're trying to determine.. but if one doesn't already know this, it's hard to understand what this code is doing.

we should try to make this code understandable to somebody that understand bash and delphix, but not necessarily how ubuntu's packages works.. so comments here would help explain the assumptions we have about ubuntu's packages. it'll also make it easier to refactor/fix this code if ubuntu breaks those assumptions.

prakashsurya commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+# Returns the default (usually latest) kernel version for a given platform.
+# Result is placed into _RET.
+#
+function get_kernel_for_platform() {
```
can you define "platform" here? I see this term used to mean different things (e.g. contrast the meaning here, with the `check_platform` function); the inconsistency can be confusing.

pzakha replied:
Perhaps I can rename `check_platform()` to make it less confusing.

prakashsurya commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+		pkg=$(echo "$line" | sed 's/^\s*//;s/\s*$//')
+		[[ -z "$pkg" ]] && continue
+		# ignore comments
+		[[ ${pkg:0:1} == "#" ]] && continue
```
we could bake this into the existing `sed` command above; `s/^#.*$//`.. doesn't really matter either way.

prakashsurya commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+		"set when BUILD_ALL=false"
+	logmust mv "packages/$SINGLE_PACKAGE_NAME/tmp/artifacts"/* artifacts/
+else
+	logmust read_package_list "$TOP/package-lists/buildall.pkgs"
```
can you comment on why we want to use a file that lists the packages, rather than dynamically generating this list using `find` (or equivalent).. i.e. generating the list of packages to build by looking at the contents of the `packages` directory..

prakashsurya commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+	unset SINGLE_PACKAGE_REVISION
+
+	for pkg in "${PACKAGES[@]}"; do
+		logmust check_valid_package "$pkg"
```
why do this again? isn't it already done in `read_package_list`?

pzakha replied:
It's redundant in this particular case. In other cases a package name can be passed from `SINGLE_PACKAGE_NAME` and requires this check, so I've decided to have it called every time we dereference a package for consistency purposes.

prakashsurya commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+		logmust ./buildpkg.sh $build_flags "$pkg"
+	done
+
+	logmust pushd metapackage
```
why don't we build `metapackage` the same way that we build other packages? i.e. using `buildpkg.sh`?

pzakha replied:
the metapackage depends on all other packages to be built first, so it doesn't make sense to call it by itself. Every package was meant to be able to be built individually.

pzakha replied:
I've added a default `all` rule to make, which calls `buildall.sh`. I've decided to leave the `metapackage` as is since IMO it really doesn't fit into the same model as the other packages

pzakha commented on Makefile:
```diff
@@ -0,0 +1,36 @@
+	shellcheck \
+	shfmtcheck
+
+clean:
```
I think we could have `make` call `./buildall.sh` since that's the default deliverable for the repository.

pzakha commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+}
+
+function install_pkgs() {
+	logmust sudo env DEBIAN_FRONTEND=noninteractive apt-get install -y "$@"
```
I don't think those would apply to the bootstrap VM since we are not changing default configuration. If this changes and some configuration changes cause this to fail, we'd investigate the failure and take a decision based on that.

prakashsurya replied:
OK, fair enough. As long as we've thought that through.

pzakha commented on setup.sh:
```diff
@@ -0,0 +1,28 @@
+
+logmust check_platform
+logmust sudo apt-get update
+logmust install_pkgs \
```
I decided not to use ansible since it adds a lot of boilerplate code (plis the requirement to have ansible installed), which I really didn't feel like it was needed given the simplicity of the setup.sh task.

pzakha commented on setup.sh:
```diff
@@ -0,0 +1,28 @@
+	debhelper \
+	devscripts
+
+logmust git config --global user.email "eng@delphix.com"
```
This makes it simpler for a user that is manually working on a previously un-configured bootstrap system to make commits. Since this framework already modifies the state of the system by installing a bunch of packages I didn't feel like adding this was more invasive.

prakashsurya replied:
IMO, there's a difference between modifying the global system state by installing packages (also something I'd rather not do if/when possible, e.g. perhaps using chroot/containers), and modifying the state of a user's home directory (which often is shared over NFS, although not in this particular case).

This sort of thing makes me think that not only should I only run the build on a bootstrap VM (which I think is a reasonable requirement, given that it's already required by other projects), but that I should have a VM used specifically for running linux-pkg and nothing else (since the approach we're taking is to clobber anything on the system as needed).

I'm not necessarily opposed to this approach, but just wanted to bring it up. I'm fine with this, if this is the design we want to go with.

pzakha replied:
So I've been mostly optimizing for easiness and speed of testing and iterating. Docker would have been a good alternative if we really wanted to make this as portable as possible at the expense of speed, however for now some packages don't build properly in docker (e.g. `bpftrace`).

prakashsurya replied:
I don't necessarily imply docker; packages have been built in isolation using chroot long before docker came around. Debian's package guidelines states the following:

> To make sure that a Debian package meets all build dependencies and is not influenced by anything specific to the user's environment, packages should be built in a chroot environment. Tools like pbuilder can be used for this.

From here: https://wiki.debian.org/HowToPackageForDebian#Building_Debian_packages

Regardless, if this is the design we want to go with, I won't object. We can always refine and iterate later if we want to (even though I doubt we will even if we decide we want to, judging by the lack of improvement to our illumos package build system).

pzakha commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+
+logmust cd "$TOP"
+
+if [[ -n "$BUILDER_CUSTOM_ENV" ]]; then
```
I will add some comments for sure.

I'm still debating if I want to add this to the documentation since this is really just a shim for Jenkins. Users should set variables explicitly instead of relying on this field.

prakashsurya replied:
Personally, I'd rather have Jenkins use the same code paths as any other "normal" user of this repository. If "users" should set variables explicitly, I vote for Jenkins doing the same (to me, it's just another user, just like any human using the repository), and removing the ability to do otherwise.

I'm fine leaving things as-is, if you disagree.

pzakha commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+
+if [[ -n "$BUILDER_CUSTOM_ENV" ]]; then
+	echo "Parsing BUILDER_CUSTOM_ENV..."
+	# TODO: add safety checks?
```
I can replace this by a WARNING instead. I am not planning to put safety checks in there for now.

prakashsurya replied:
Generally speaking, I also don't see much use in "warning" output lines unless they're actionable (e.g. compiler warnings). In this case, the warnings will be ignored (by design, since this is a feature), so I don't see much value it them.

pzakha commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+	export KERNEL_VERSIONS
+fi
+
+if [[ -n "$SINGLE_PACKAGE_NAME" ]]; then
```
I agree that it can be confusing.
So the intention here is to have `buildall.sh` be the entry script from jenkins. An alternative here would be to create a wrapper script which would be called by jenkins instead and have it call `buildall.sh` or `buildpkg.sh` depending on the parameters passed. This might be a better way to handle this.
The idea of having SINGLE_PACKAGE_NAME is be able to still build all the packages while having some changes for one particular package, so that the resulting package bundle can be consumed by appliance-build later.
Thinking of it I'm leaning more and more into creating an additional wrapper as it would might cleanup this code significantly.

pzakha commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+	logmust install_pkgs $headers_pkgs
+}
+
+function default_revision() {
```
I've put `delphix` to make it really explicit that the package is generated by Delphix.
That said this is just a temporary value since we needed to put something here. We'll need to decide on a versioning schema some time soon at which point we'll be able to set this appropriately.

prakashsurya replied:
I think the actual value we're using here is good, I just want to have these things documented.

E.g. add a comment to state that "delphix" is used in the version so we can use simple tools like `dpkg -l | grep delphix` to get a list of all delphix built packages on the system.

Also, we can add a comment to state we just need the package revisions to be increasing, so things like "apt-get dist-upgrade" pull in the "new" packages automatically, which lends the time of the build as a good way to ensure this "increasing" property.

I doubt we'll go back and change this, perhaps except if we want this to match the version of the virtualization package and/or "release version" (e.g. 5.3.0.0).

pzakha commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+		#
+		# Check that the target kernel is valid
+		#
+		apt-cache show "linux-image-${kernel}" >/dev/null 2>&1 ||
```
Mostly because I want to suppress the default output of `apt-cache show`

prakashsurya replied:
if `logmust` outputed its lines to stderr, we could do this:
```
logmust apt-cache show ... >/dev/null
```
but, it's not a big deal.. I'm fine with this as-is.

pzakha commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+	#
+	if [[ "$(apt-cache show --no-all-versions "linux-image-${platform}" \
+		2>/dev/null | grep Depends)" =~ linux-image-([^,]*-${platform}) ]]; then
+		_RET=${BASH_REMATCH[1]}
```
Yeah, I had other commands in the script that used `_RET` which I removed subsequently. So we could use echo, but we need to be super careful to make sure other commands don't echo anything. I feel like `_RET` is a more robust approach and would like to see it standardized in more areas.

pzakha commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash -e
```
I have put it there in case I've missed checks in some places. I think it's better than the default of not having it. I don't feel too strongly about it though.

prakashsurya replied:
can we use `-e` and the convetion of `|| die` and/or `logmust` at the same time?

if `-e` is set, then we'll never actually call `die` after the first command fails, right? or maybe I misunderstand?

pzakha replied:
We can, however there is another case where it could cause an issue (e.g. if we execute something and then check for value of `$?). We don't have this kind of check in the code, but that might cause an unexpected issue if we do. I'm considering removing it.

Here's the bash docs for `set -e`:
```
  -e      Exit immediately if a simple command (see SHELL GRAMMAR above)  exits
          with  a non-zero status.  The shell does not exit if the command that
          fails is part of the command list immediately following  a  while  or
          until  keyword,  part of the test in an if statement, part of a && or
          || list, or if the command's return value is being inverted via !.  A
          trap on ERR, if set, is executed before the shell exits.
```

prakashsurya replied:
interesting! OK, thanks. I wasn't aware of that.

pzakha commented on buildall.sh:
```diff
@@ -0,0 +1,127 @@
+# limitations under the License.
+#
+
+TOP="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null && pwd)"
```
Good idea. It's hard to keep track what works on each platform.

prakashsurya commented on packages/delphix-platform/config.sh:
```diff
@@ -0,0 +1,50 @@
+
+function checkstyle() {
+	logmust cd "$WORKDIR/repo"
+	logmust make check
```
this isn't really "correct" .. the platform package provides a Dockerfile so we can have a consistent environment for running the make targets, which you're not using here.

I understand that you don't want to rely on docker, but for delphix-platform, the assumption is docker is required and used. We even use it in the current build via jenkins:
```
109         stage('Build') {                                                        
110           dir('delphix-platform') {                                             
111             ansiColor('xterm') { sh './scripts/docker-run.sh make package' }    
112           }                                                                     
113         }  
```

Since this is just another consumer of the delphix-platform repository, it's free to do whatever it wants.. but as one of the maintainers of delphix-platform, I don't want to support/maintain running these make targets outside of the Docker image/environment. The point of the Dockerfile is so we can have a consistent runtime environment no matter where these targets are executed, and that environment can evolve in lockstep with the code that runs in it (e.g. if we need to add/change the environment, there's no flag days to worry about).

pzakha replied:
I could modify the build step in `linux-pkg` to use the make targets in `delphix-platform` rather than the current method.
I can modify the Makefile to ingest `PACKAGE_VERSION` and `PACKAGE_REVISION` environment variables when they are defined.

I'm wondering though if that complexity is necessary. We can use the following command to install build dependencies from a control file, so we don't need to set them explicitly in the `prepare()` hook if that's one of your concerns:
```
sudo mk-build-deps --install --remove debian/control
```

As for having a consistent build environment, do you have any particular worries? For the time being `delphix-platform` has integration with bors & travis, which is a valid reason for leaving all this bootstrap & docker code around, however if we had automated testing of PRs integrated with the `linux-pkg` framework, would we still need all that?

prakashsurya replied:
> I can modify the Makefile to ingest PACKAGE_VERSION and PACKAGE_REVISION environment variables when they are defined.

Yea, I'd be OK with such a change.

> however if we had automated testing of PRs integrated with the linux-pkg framework, would we still need all that?

You're suggesting that as part of the automated testing of PRs to the delphix-platform project, we pull in the linux-pkg repository to test the delphix-platform change? If so, I don't think that solution is better than what we're currently doing for delphix-platform (e.g. flag days). I don't think it's even possible (running linux-pkg using TravisCI), which is what we're currently using to run the tests against each PR/commit.

Or are you suggesting we rely on PRs to the linux-platform project, for testing changes to delphix-platform? In that case, we'd completely lose all automated testing of the delphix-platform PRs and commits, right? That also doesn't sounds better than what we currently have. This is also something that doesn't yet exist (automated tests of the linux-pkg PRs and commits), so it'd be a regression until we implemented it.

> As for having a consistent build environment, do you have any particular worries?

I want to be certain, for every commit that goes into delphix-platform, the following are true:

1. `make package` works
2. `make shfmtcheck` works
3. `make shellcheck` works

Since all of those commands depend on being run in a specific environment, I'm defining "works" to mean "works when run in the environment defined by the Dockerfile".

I do this because I want to make sure that changes to the project don't break users of the project (ideally we'd also do functional tests), and using docker is the best way I can ensure the environment used by the automated tests is the same as what a user will use.

If we don't want to use docker in linux-pkg, that's fine, but the delphix-platform project makes no guarantees that what goes into it will not break linux-pkg; since the environment linux-pkg uses to build may (or may not) be different than what was verified to work by the automated tests.

Using `mk-build-deps` to pull in build dependencies would be good, but it's still not sufficient since that's not what the automated PR/commit tests will do; so it's not a workflow that will have been verified to work. Further, since TravisCI doesn't enable us to provide a specific build environment to run in (except via docker), we can't simulate the "bootstrap-18-04" environment that linux-pkg wants to run in.

Ultimately, I just want to ensure every commit to delphix-platform gets tested and is verified to work so as not to break consumers of the project (e.g. linux-pkg). If we could tie Jenkins up to delphix-platform, and have a job that used the "bootstrap-18-04" image to do the automated tests, and did the same things that linux-pkg wants to do, then I'd feel more comfortable not using docker to do the build in linux-pkg.

I hope that makes sense; let me know if I should clarify on anything. I'm not trying to be difficult, I just want things to be automatically tested and for users to use the tested code paths, in an attempt to avoid accidental breakage.

prakashsurya commented on packages/delphix-platform/config.sh:
```diff
@@ -0,0 +1,50 @@
+		logmust git reset --hard repo-HEAD
+		logmust git clean -qdxf
+		logmust ./configure.sh
+		logmust dpkg_buildpackage_default
```
This project already has a way to build the package, which we're not using here.

I don't really want to maintain something other than the current way of building the package, which is `make package` running in the docker container. I say this because I know for sure that will always work since it's tested and enforced on each commit to the repository, using the automatic PR and commit testing via TravisCI and bors.

The tight coupling between the delphix-platform repository, and this repository worries me. It makes automatic testing much more difficult, and opens the door to lots of flag days that don't currently exist in the current build system. To me, this feels like a regression, at least for delphix-platform in particular.

prakashsurya commented on README.md:
```diff
@@ -0,0 +1,736 @@
+
+TODO
+
+## Versions and Branches
```
I have a couple high level questions...

1. Do we plan to be able to rebuild old versions of packages from source? or rely on saving/reusing previously built artifacts?

I think the answer is we do want to be able to rebuild old versions of packages, but I wanted to ask to be sure.

To achieve this, it means we need to store the source for each package, and never just pull the "latest" version of upstream and build from that. I think this is how it works right now, but please correct me if I'm wrong.

It also means we need to ensure we can replicate an old build environment previously used to build all the packages (or something equivalent). Right now, since we just install the "latest" build dependencies from ubuntu, I don't think we can replicate an old build environment, but hopefully it'll be close enough to allow us to build mostly equivalent packages. If we use our internal APT mirrors instead of pulling from Ubuntu, this might practically solve the problem.

Replicating an old build environment has been a big problem for us in the past, and we've just been lucky enough to never have a build server fail on us.

2. Will we be able to change a single package (e.g. cloud-init), and install only that package via a hotfix and/or upgrade image?

As an example, if we need to fix cloud-init.. can we make that fix, and ensure only the cloud-init package (and maybe delphix-entire) is changed on the customers system? How will we go about making the change, building the package, building the hotfix/upgrade image, delivering the hotfix/upgrade image, and then ultimately applying the hotfix/upgrade image to the customer's system? I understand all the work to execute this workflow might not be implemented yet, but I just want to understand if we have a plan to implement it, and what that plan might be. If we don't have a plan for this yet, that's fine too.

I might have overlooked where this question is already answered, so let me know if it's already explained somewhere.

pzakha replied:
For 1), yes we do want to be able to rebuild old packages from source. I've talked with Basil and we are planning to fully mirror Ubuntu archives (and custom PPAs) in our local mirrors, including source packages. The goal will be to configure the package lists during linux-pkg builds such as we only use our mirrors.

For 2) it's unclear for now how we will be handling hotfixes. Using cloud-init as an example, let's say we want to hotfix 5.4.1.1 with a custom version of cloud-init. We could create a new branch in `linux-pkg` based on 5.4.1.1, say `5.4.1.1-HF3099`. In that new branch we could provide an alternative version/revision for cloud-init, so calling `buildall.sh` would produce the same debs as we had before for all other packages except for cloud-init. We would then have to build a new upgrade image using the S3 artifacts produced from linux-pkg branch `5.4.1.1-HF3099`.

prakashsurya replied:
> I've talked with Basil and we are planning to fully mirror Ubuntu archives (and custom PPAs) in our local mirrors, including source packages.

We'll need separate mirrors for each old release (including maintenance and patch releases), right? Basically, anything that we want to rebuild (and/or create a hotfix from), will need a unique mirror with contents that don't change?

Otherwise, as an example, we could update the version of a compiler (or other build dependency) in our mirror, which could potentially change the contents of a binary package, even if the source code for that package didn't change; right?

> We would then have to build a new upgrade image using the S3 artifacts produced from linux-pkg branch 5.4.1.1-HF3099.

If we use the normal upgrade logic to apply the change, I think only the changed package will be updated when doing an in-place upgrade; assuming the versions of all other packages don't change, apt-get won't update/change them.

If we do a not-in-place upgrade to deploy the fix, then I think we'll end up installing all the other packages as well. This probably is fine, but since we don't have bit-for-bit reproducible builds, it does mean we'll be installing new/different binaries for packages we didn't intend to change.

I'm not trying to imply any of this is bad/wrong, I'm just trying to understand all of the potential implications as best as I can.

pzakha replied:
> We'll need separate mirrors for each old release (including maintenance and patch releases), right? Basically, anything that we want to rebuild (and/or create a hotfix from), will need a unique mirror with contents that don't change?

Yeah, we will have distinct mirrors for each release and staging branch.

> If we do a not-in-place upgrade to deploy the fix, then I think we'll end up installing all the other packages as well. This probably is fine, but since we don't have bit-for-bit reproducible builds, it does mean we'll be installing new/different binaries for packages we didn't intend to change.

If we do a not-in-place upgrade then yes. I think having bit-for-bit reproducible bits shouldn't be an issue. As long as we use the same source as before that shouldn't cause any issues. If we produce functionally different packages in 2 different builds using the same source, then we have a much bigger problem.

pzakha commented on lib/common.sh:
```diff
@@ -0,0 +1,560 @@
+	local var
+
+	prefix=$(echo "$PACKAGE" | tr - _ | tr '[:lower:]' '[:upper:]')
+	[[ "$prefix" =~ ^[A-Z][A-Z0-9_]*$ ]] ||
```
`check_valid_package` makes sure that we are passing a package known by the framework, so it doesn't need to do this check. I'll create a new function `load_package_config()` that will be responsible of validating the package name and the package config.

pzakha commented on buildpkg.sh:
```diff
@@ -0,0 +1,263 @@
+	logmust touch "$WORKDIR/repo-updated"
+
+	logmust cd "$WORKDIR"
+	return 0
```
In most cases it's not necessary. It would only be necessary if there is an unchecked test right before (`[[ $var == $value ]]`), or the command before it failed, and both of those would be problematic.

pzakha commented on buildpkg.sh:
```diff
@@ -0,0 +1,263 @@
+	logmust git fetch upstream "$UPSTREAM_GIT_BRANCH"
+
+	logmust git branch repo-HEAD FETCH_HEAD
+	logmust git branch upstream-HEAD FETCH_HEAD
```
I think it's cleaner to have `repo-HEAD` and `upstream-HEAD`, this way other scripts don't need to care what exact branch we have fetched.

pzakha commented:
I've updated the review, with most of the feedback addressed.
Here is the summary of the changes:
 - Add delphix-kernel as a separate package. (See https://github.com/delphix/delphix-kernel/pull/1)
 - delphix-platform is now built using its Makefile. (Updated PR for delphix platform changes: https://github.com/delphix/delphix-platform/pull/17)
 - Add new wrapper script jenkins-build.sh and simplify buildall.sh.
 - Add a template config.sh file.
 - Add function load_package_config that loads config.sh and performs checks on it.
 - Add internal link for temporary testing instructions.
 - Refactor some environment variables logic to make it cleaner.
 - Updated README and added link to temporary test instructions

pzakha commented on jenkins-build.sh:
```diff
@@ -0,0 +1,158 @@
+#!/bin/bash
```
So one easy solution would be to have `jenkins_build.sh` execute `make all` and `make <package>` instead of `updateall.sh` and `buildpkg.sh`. I'll just need to have `buildpkg.sh` consume the `CHECKSTYLE` environment variable.

Manual users would be free to use `make all`, `make <package>` or to run `./buildpkg.sh <custom flags> <package>` depending on their needs.

sebroy commented on README.md:
```diff
@@ -0,0 +1,758 @@
+
+* **TARGET_PLATFORMS**: Some packages build kernel modules. This specifies which
+  kernel versions to build those packages for and accepts a space-separated list
+  of values. If the value is a platform, such as "aws" or "generic", then it
```
Curious: what is "generic" used for? Also, where is the complete list of supported platforms that can be passed in via this variable?

sebroy commented on README.md:
```diff
@@ -0,0 +1,758 @@
+
+#### Creating debian metadirectory
+
+TODO
```
Do you intend to fill this in before pushing?

pzakha replied:
I wasn't intending to. Perhaps I can leave a link to https://www.debian.org/doc/manuals/maint-guide/dreq.en.html for now?
